src/compiler/commandLineParser.ts

```diff
@@ -800,6 +800,15 @@ const commandOptionsWithoutBuild: CommandLineOption[] = [
         description: Diagnostics.Raise_an_error_when_a_function_parameter_isn_t_read,
         defaultValueDescription: false,
     },
+    {
+        name: "noUnnecessaryTypeAssertions",
+        type: "boolean",
+        affectsSemanticDiagnostics: true,
+        affectsBuildInfo: true,
+        category: Diagnostics.Type_Checking,
+        description: Diagnostics.Raise_an_error_when_a_type_assertion_does_not_affect_the_type_of_an_expression,
+        defaultValueDescription: false,
+    },

```

Comment:
Oooh boy. The last I remember hearing of "linting" style features in TypeScript around `noUnusedLocals` and `noUnusedParameters` was that the team was strongly considering moving away from them? _(this comment should not be treated as a source - I'm asking about anecdotal memories)_

As much as I _love_ making this a feature, we've already come into conflict with the `noUnused*` flags in lint land with https://typescript-eslint.io/rules/no-unused-vars. Users often want more flexibility than TypeScript can offer with just compiler options. See https://eslint.org/docs/latest/rules/no-unused-vars#options for options demanded by enough users to be justified adding to ESLint core, and https://github.com/typescript-eslint/typescript-eslint/issues/5271 for a reference on many of the wacky ways users have asked to be able to configure rules.

Is there more context that can be provided on the TypeScript team's opinion on adding more compiler options for lint-area responsibilities? Perhaps a better long-term strategy would be to provide an API we can hook into? Perhaps even a ... **[type relationship API](https://github.com/microsoft/TypeScript/issues/9879)**? 😁

Reply:
My impetus for suggesting this was #51456, where I was trying to stop type checking during lint as doing so is half of our lint time, which is not so good for editor responsiveness in the project (for those of us who run ESLint in the editor; I think I may be one of the only ones on the team willing to take the perf hit of running it as an extension on this repo).

We only wanted this one rule, which is something that TS could be doing internally pretty easily.

But, I'm not really the one to be answering for the whole team's opinion here.

Reply:
If linting with type info is slow for you guys, perhaps we can work together to figure out how we can improve the performance overall?

Obviously we won't be able to get it to the same speed as non-type-aware lint, but perhaps we can close that gap. This would ofc be a great win for the entire community!

Reply:
I'd love to come up with ways to do so, though given the time is roughly the time it takes to compile TS itself, it's hard to know what could be improved besides improving the performance of TS itself. (e.g., upgrade to 5.0 😄)

Reply:
I mean, it's basically slow because it's duplicating all the work the compiler already does in the language service to fetch errors in a file for the editor, or at least a large part of that work, because calculating the types at a bunch of locations is most of the work the compiler does. I also don't think the eslint editor extension even uses our language service APIs to keep around a persistent project, so it's also not incremental in its parse behavior as far as the TS API is concerned (even though normal lints may be incremental).

Honestly, I'd still love to see linter plugins that could be integrated straight into the compiler at some point, like I implemented back when I was an intern - being able to guarantee a single walk that's incremental, shared across all rules, and reuses checker data where possible would be big cross-project upside, we just are still skittish about a built-in plug-in model after all this time because we're scared both about untrusted code in the editor (maybe passable now that the editor asks if you trust the code) and about being blamed for the performance penalty poorly written plugins may cause (since perf is such a huge concern).

Reply:
> Is there more context that can be provided on the TypeScript team's opinion on adding more compiler options for lint-area responsibilities?

What we've said on this, specifically, in our discussions at this point is that while we prefer things like this to be external lint rules, if there's outsized cost or capability lost by them being external (eg, because they require fancy control flow checks to properly implement), we're more OK with them living in the compiler for now. So current internal consensus is mostly that stuff like `noFallthroughInSwitch` really should have been external, since it's a fast syntactic check, `noUnusedX` is a bit more middling since it's kinda free for us to track while already checking, and stuff like this that needs control flow based definite assignment analysis and relationship checks to answer right is right up the alley of what our quickfixes should pick up, since we're very well-positioned to answer this question quickly during a normal type check.

Sans the compiler option to make this into an error, this check & quick fix is almost a carbon copy of our unneeded `await` quickfix, so we've had similar functionally knocking around for awhile.

Reply:
> eslint editor extension even uses our language service APIs

cc @bradzacher - is this something that ESLint today could support?

Most of typescript-eslint is pluggable and supports receiving TypeScript programs. For example, the parser's options can include a `program`: https://github.com/typescript-eslint/typescript-eslint/blob/07119945fb0c1d6acb421b1b9ba972c8d9c8942a/packages/types/src/parser-options.ts#L53.

This lack of language service / program reuse is one of our biggest pain points. We added an [FAQ about out-of-date type info](https://typescript-eslint.io/docs/linting/troubleshooting#changes-to-one-file-are-not-reflected-when-linting-other-files-in-my-ide) to our site. https://github.com/eslint/eslint/discussions/16557#discussioncomment-4160219 has context on the ESLint structural side.

> linter plugins that could be integrated straight into the compiler at some point

+1. https://github.com/Quramy/typescript-eslint-language-service is the closest I've seen used in the wild.

> outsized cost or capability lost by them being external

Separate from the performance concerns, `no-unnecessary-type-assertion` is one of the less costly rules in TypeScript-ESLint core. It's only a couple hundred lines of real logic. [Searching for accepted `no-unnecessary-type-assertion` issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=no-unnecessary-type-assertion+in%3Atitle+is%3Aissue+-reason%3A%22not+planned%22+-label%3Awontfix+-label%3Aexternal+-label%3A%22fix%3A+user+error%22) shows only 22 open since the beginning of 2019. [Only 13 of those are bugs](https://github.com/typescript-eslint/typescript-eslint/issues?q=no-unnecessary-type-assertion+in%3Atitle+is%3Aissue+-reason%3A%22not+planned%22+-label%3Awontfix+-label%3Aexternal+-label%3A%22fix%3A+user+error%22+label%3Abug). I'd put forward that the maintenance cost on our side for it is relatively quite small.

Reply:
> is this something that ESLint today could support?

No, we don't know when we're running in the IDE, so we can't spin up a server.

But we default to using a `ts.WatchCompilerHostOfConfigFile` (and by extension a `ts.BuilderProgram`), so we are doing incremental builds using persistent programs.
This is also cached forever and reused whenever a file that is included in the program is linted.